Add `backup` and `restore` script for bootstrap disaster recovery

TeddyAndrieux opened the pull request:
DRAFT PR: Currently not working

**Component**:

'salt', 'kubernetes'
<!-- E.g. 'salt', 'containers', 'kubernetes', 'build', 'tests'... -->

**Context**: 

Add some backup and be able to recovery a bootstrap node

**Summary**:

Add two script for backup and restore a bootstrap

**Procedure**:
Backup:
- Automaticaly done at the end of the bootstrap
- to create a new backup just run `backup.sh` script
- by default all backup are stored in `/var/lib/metalk8s`

Restore:
- Copy a backup archive on a new host
- Get all ISOs at the same place as the previous bootstrap node
- Mount one ISO and run the restore script with `-b <backup_file_path>`
- /!\ If you have a 3 member etcd cluster then you will need to remove the old bootstrap etcd member before running the restore script (https://github.com/etcd-io/etcd/blob/34fcabab55585b39835820278da4de65b3463358/Documentation/faq.md#why-wont-etcd-accept-my-membership-changes)
- /!\ restore is impossible if you do not have an HA apiserver because we need apiserver to get ssh informations to reconfigure all salt-minions

**Acceptance criteria**: 

Be able to restore a bootstrap node from a previous backup

---

Fixes: #1615

TeddyAndrieux committed:
Add `backup` and `restore` script for bootstrap disaster recovery

Fixes: #1615

bert-e commented:
# Hello teddyandrieux,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @teddyandrieux

* @monpote

* @jbertran

* @sayf-eddine-scality

* @nicolast

* @cmonfort

* @slaperche-scality

* @thomasdanan

* @zempashi

* @lucieleonard

* @gdemonet

* @carlito-scality

* @nootal

* @alexandre-allard-scality

TeddyAndrieux committed:
Include m2crypto in apiserver kubeconfig

kubeconfig state require m2crypto state but the sls is not included so
not able to run directly the kubeconfig state.

TeddyAndrieux committed:
Add `list` support for kubernetes roster

TeddyAndrieux committed:
Only filter on minion already in the mine when getting etcd endpoints

One etcd salt-minion may not be available when trying to get etcd
endpoints up

bert-e commented:
# Conflict




There is a conflict between your branch `improvement/1615-bootstrap-backup-recovery` and the
destination branch `development/2.4`.

Please resolve the conflict on **the feature branch** (`improvement/1615-bootstrap-backup-recovery`).

```bash
 $ git fetch
 $ git checkout origin/improvement/1615-bootstrap-backup-recovery
 $ git merge origin/development/2.4
 $ # <intense conflict resolution>
 $ git commit
 $ git push origin HEAD:improvement/1615-bootstrap-backup-recovery
```

gdemonet commented on scripts/backup.sh:
```diff
@@ -0,0 +1,203 @@
+
+run "Backing up MetalK8s configurations" backup_metalk8s_conf
+run "Backing up CAs certificates and keys" backup_cas
+run "Backing up ETCD" backup_etcd
```
Not sure I ever saw `etcd` capitalized like this, maybe something like `Backing up etcd data` would make sense.

gdemonet commented on scripts/backup.sh:
```diff
@@ -0,0 +1,203 @@
+    done
+}
+
+backup_etcd() {
```
~So, IIUC, this method only does a "local" snapshot of the `etcd` data, as seen by the Bootstrap member. That's perfectly fine if we have a 1-node `etcd` cluster, but not that useful since we rely on API Server (hence, on the availability of an `etcd` cluster) to restore such a cluster.~

~OTOH, if we have a larger `etcd` cluster, do we need to restore from a backup? Wouldn't it be better to snapshot an active member and restore the new Bootstrap from this snapshot? It would at least allow for more recent data.~

Answer was already [in the ticket](https://github.com/scality/metalk8s/issues/1615#issuecomment-528915235), snapshot is not used in the restore script.

gdemonet commented on scripts/restore.sh.in:
```diff
@@ -0,0 +1,442 @@
+
+restore_cas() {
+    local -r ca_dir='/etc/kubernetes/pki/'
+    local -a ca_files=(
```
What if the backup was made in some different version? Maybe we could simply restore whatever is available in `${BACKUP_DIR}/pki`, wouldn't that be acceptable?

gdemonet reviewed (dismissed):
Code LGTM, will test the `restore.sh` script on an active Control-Plane Node (as specified in the issue) to ensure it's feasible. Great effort, we can feel much safer now :+1: :1st_place_medal: !

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @teddyandrieux

* @monpote

* @jbertran

* @sayf-eddine-scality

* @nicolast

* @cmonfort

* @slaperche-scality

* @thomasdanan

* @zempashi

* @lucieleonard

* @gdemonet

* @carlito-scality

* @nootal

* @alexandre-allard-scality

sayf-eddine-scality commented:
LGTM and tested
```
NAME        STATUS     ROLES                         AGE   VERSION
bootstrap   NotReady   bootstrap,etcd,infra,master   89m   v1.15.3
centos1     Ready      etcd,master,node              83m   v1.15.3
centos2     Ready      bootstrap,etcd,infra,master   11m   v1.15.3
centos3     Ready      etcd,master,node              58m   v1.15.3
```
one thing the backup in the end of the restore script does not wait for the etcd to be up and running and thus crashes

TeddyAndrieux committed:
Check etcd cluster health after register

TeddyAndrieux commented:
```
[root@bootstrap02 ~]# /srv/scality/metalk8s-2.4.0-dev/restore.sh -b backup_20190913_194404.tar.gz 
> Extract backup archive 'backup_20190913_194404.tar.gz'... done [0s]
> Pre-minion system tests... done [0s]
> Disabling Salt minion service... done [0s]
> Stopping Salt minion service... done [0s]
> Configuring local YUM repositories... done [0s]
> Installing Salt minion... done [8s]
> Installing genisoimage... done [1s]
> Configuring Salt minion to run in local mode... done [5s]
> Restoring MetalK8s configurations... done [2s]
> Restoring CAs certificates and keys... done [0s]
> Configuring salt-master... done [217s]
> Pushing CAs to salt mine... done [10s]
> Marking new bootstrap node... done [8s]
> Reconfiguring all nodes... done [51s]
> Applying highstate on the new bootstrap node... done [134s]
> Backing up MetalK8s configurations... done [0s]
> Backing up CAs certificates and keys... done [0s]
> Backing up etcd data... done [0s]
> Creating backup archive '/var/lib/metalk8s/backup_20190913_212201.tar.gz'... done [0s]
```

TeddyAndrieux commented:
/approve

slaperche-scality commented:
piratos :grin:

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.4`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.4`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



Please check the status of the associated issue None.

Goodbye teddyandrieux.
